File: 12/gpt-5.1/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func parseInput(path string) (map[int]int, []string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, nil, err
	}
	defer file.Close()

	shapeSizes := make(map[int]int)
	var regionLines []string
	currentShapeId := -1
	currentShapeCount := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimRight(scanner.Text(), "\r")
		if strings.TrimSpace(line) == "" {
			continue
		}

		if len(line) > 0 && line[0] >= '0' && line[0] <= '9' && strings.Contains(line, ":") && !strings.Contains(line, "x") {
			if currentShapeId >= 0 {
				shapeSizes[currentShapeId] = currentShapeCount
			}
			idPart := strings.SplitN(line, ":", 2)[0]
			idVal, err := strconv.Atoi(idPart)
			if err != nil {
				return nil, nil, err
			}
			currentShapeId = idVal
			currentShapeCount = 0
		} else if strings.Contains(line, "x") && strings.Contains(line, ":") {
			if currentShapeId >= 0 {
				shapeSizes[currentShapeId] = currentShapeCount
				currentShapeId = -1
				currentShapeCount = 0
			}
			regionLines = append(regionLines, strings.TrimSpace(line))
		} else if currentShapeId >= 0 && (strings.Contains(line, "#") || strings.Contains(line, ".")) {
			currentShapeCount += strings.Count(line, "#")
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, nil, err
	}

	if currentShapeId >= 0 {
		shapeSizes[currentShapeId] = currentShapeCount
	}

	return shapeSizes, regionLines, nil
}

func solvePart1(path string) (int, error) {
	shapeSizes, regionLines, err := parseInput(path)
	if err != nil {
		return 0, err
	}

	countFit := 0
	for _, regionLine := range regionLines {
		parts := strings.SplitN(regionLine, ":", 2)
		if len(parts) != 2 {
			continue
		}
		sizePart := strings.TrimSpace(parts[0])
		countsPart := strings.TrimSpace(parts[1])

		sz := strings.Split(sizePart, "x")
		if len(sz) != 2 {
			continue
		}
		w, _ := strconv.Atoi(sz[0])
		h, _ := strconv.Atoi(sz[1])

		countStrs := strings.Fields(countsPart)
		totalCells := 0
		for i, s := range countStrs {
			c, _ := strconv.Atoi(s)
			if c == 0 {
				continue
			}
			if size, ok := shapeSizes[i]; ok {
				totalCells += c * size
			}
		}

		gridArea := w * h
		if totalCells <= gridArea {
			countFit++
		}
	}

	return countFit, nil
}

func main() {
	part1, err := solvePart1("input.txt")
	if err != nil {
		panic(err)
	}

	fmt.Printf("Day 12 Part 1: %d\n", part1)

	// Part 2 placeholder
	var part2 int64 = 0
	fmt.Printf("Day 12 Part 2: %d\n", part2)
}

File: 12/gpt-5.1/rust/src/main.rs
use std::collections::HashMap;
use std::fs;

fn parse_input(path: &str) -> (HashMap<usize, usize>, Vec<String>) {
    let content = fs::read_to_string(path).expect("Failed to read input");
    let lines: Vec<String> = content
        .lines()
        .map(|l| l.trim_end_matches('\r').to_string())
        .collect();

    let mut shape_sizes: HashMap<usize, usize> = HashMap::new();
    let mut region_lines: Vec<String> = Vec::new();
    let mut current_shape_id: Option<usize> = None;
    let mut current_shape_count: usize = 0;

    for line in lines {
        if line.trim().is_empty() {
            continue;
        }

        let is_shape_header = line.chars().next().map(|c| c.is_ascii_digit()).unwrap_or(false)
            && line.contains(':')
            && !line.contains('x');

        if is_shape_header {
            if let Some(id) = current_shape_id.take() {
                shape_sizes.insert(id, current_shape_count);
            }
            let id_part = line.split(':').next().unwrap();
            let id: usize = id_part.parse().expect("invalid shape id");
            current_shape_id = Some(id);
            current_shape_count = 0;
        } else if line.contains('x') && line.contains(':') {
            if let Some(id) = current_shape_id.take() {
                shape_sizes.insert(id, current_shape_count);
                current_shape_count = 0;
            }
            region_lines.push(line.trim().to_string());
        } else if current_shape_id.is_some() && (line.contains('#') || line.contains('.')) {
            current_shape_count += line.chars().filter(|&c| c == '#').count();
        }
    }

    if let Some(id) = current_shape_id {
        shape_sizes.insert(id, current_shape_count);
    }

    (shape_sizes, region_lines)
}

fn solve_part1(path: &str) -> i64 {
    let (shape_sizes, region_lines) = parse_input(path);
    let mut count_fit: i64 = 0;

    for region_line in region_lines {
        let mut parts = region_line.splitn(2, ':');
        let size_part = parts.next().unwrap().trim();
        let counts_part = parts.next().unwrap_or("").trim();

        let mut size_iter = size_part.split('x');
        let width: i64 = size_iter.next().unwrap().parse().unwrap();
        let height: i64 = size_iter.next().unwrap().parse().unwrap();

        let mut total_cells: i64 = 0;
        for (idx, tok) in counts_part.split_whitespace().enumerate() {
            let c: i64 = tok.parse().unwrap();
            if c == 0 {
                continue;
            }
            if let Some(size) = shape_sizes.get(&idx) {
                total_cells += c * (*size as i64);
            }
        }

        let grid_area = width * height;
        if total_cells <= grid_area {
            count_fit += 1;
        }
    }

    count_fit
}

fn main() {
    let part1 = solve_part1("input.txt");
    println!("Day 12 Part 1: {}", part1);

    // Part 2 placeholder
    let part2: i64 = 0;
    println!("Day 12 Part 2: {}", part2);
}
